consensus/core/src/dag_state.rs

```diff
@@ -83,27 +87,355 @@ impl DagState {
         self.cached_refs[block_ref.author].insert(block_ref);
     }
 
-    /// Gets a copy of an uncommitted block. Returns None if not found.
-    /// Uncommitted must be in memory, so only in-memory blocks are checked.
+    /// Gets an uncommitted block. Returns None if not found.
+    /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
     pub(crate) fn get_uncommitted_block(&self, reference: &BlockRef) -> Option<VerifiedBlock> {
         self.recent_blocks.get(reference).cloned()
     }
 
-    pub(crate) fn get_blocks_at_slot(&self, _slot: Slot) -> Vec<VerifiedBlock> {
-        unimplemented!()
+    /// Gets all uncommitted blocks in a slot.
+    /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
+    pub(crate) fn get_uncommitted_blocks_at_slot(&self, slot: Slot) -> Vec<VerifiedBlock> {
+        let mut blocks = vec![];
+        for (block_ref, block) in self.recent_blocks.range((
```

Comment:
aah nice I see what you meant about not having to iterate over all blocks, I should have used range!